### DISCUSSION TITLE
A guideline for new coin support

### DISCUSSION THREAD

**Discussion by calvintam236** [coinfoundry/miningcore#862] [Q&A] [answered]
@oliverw can you make us a basic guideline to implement coin support, e.g. where and what to edit? i think most people have no clue to begin with, especially there isn't much comment inside code to help understanding the system structure and workflow. in addition, how about provide a template for new coin?

**Comment by Mer-idium** [comment]
@calvintam236 I believe these are the files that need to be changed. it is a simple process if the pool already supports the coins algorithm.

ClusterConfig.cs
CoinMataData.cs
BitcoinPayoutHandler.cs
BitcoinPool.cs
BitcoinProperties.cs

**Comment by oliverw** [comment, marked as answer]
@calvintam236 @pfeiflelukas I'm afraid its not that easy. At least not in general. Below is a list in order of increasing complexity:

1. The coins that are the easist to implement are Bitcoin forks with a slightly changed Proof-of-Work algorithm (examples Litecoin, Vertcoin)
2. Then we have coins that are Bitcoin forks but require some special treatment of their block template in order to satisfy consensus rules (example Dash)
3. There's another category of Bitcoin forked coins that not only come with changed PoW algorithm and consensus rule requirements but also require a completely reworked stratum implementation for the miners (examples ZCash, Bitcoin Gold)
4. The last category consists of coins that are not derived of an existing coin at all and require the implementation of the full stack of components: Pool (stratum), Job Manager, Job and Metadata.  Those are by far the most complex ones (examples Monero, Ethereum Family)

**Comment by oliverw** [comment]
Closing this as there will never be a general rule for supporting new coins. Anyone attempting to add a new coin needs to be familiar with the coin's codebase to make educated descisions on how to integrate them.
